src/shogun/mathematics/linalg/linalgRefactor.h

```diff
@@ -0,0 +1,171 @@
+ *
+ * Copyright (c) 2016, Shogun-Toolbox e.V. <shogun-team@shogun-toolbox.org>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Pan Deng
+ */
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/mathematics/eigen3.h>
+#include <iostream>
+
+using namespace shogun;
+
+template <class T>
+struct BaseVector
+{
+    BaseVector(){}
+    BaseVector(const SGVector<T> vector){}
+
+    virtual bool onGPU()
+    {
+        return false;
+    }
+};
+
+template <class T>
+struct CPU_Vector : public BaseVector<T>
+{
+    //<SGVector<T>>* CPUptr;
+    SGVector<T> vec;
+
+    CPU_Vector(const SGVector<T> vector)
+    {
+        //CPUptr = &vector;
+        vec = vector;
+    }
+
+    bool onGPU()
+    {
+        return false;
+    }
+};
+
+template <typename T>
+struct GPU_Vector : public BaseVector<T>
+{
+#ifdef HAVE_VIENNACL
+    // unique_pointer<VCLMemoryArray> GPUptr;
+    // other gpu related stuff
+#endif
+    bool onGPU()
+    {
+        return true;
+    }
+};
+
+class CPUBackend
+{
+public:
+    template <typename T>
+    T dot(CPU_Vector<T> a, CPU_Vector<T> b)
+    {
+        typedef Eigen::Matrix<T, Eigen::Dynamic, 1> VectorXt;
+        Eigen::Map<VectorXt> vec_a = a.vec;
+        Eigen::Map<VectorXt> vec_b = b.vec;
+        return vec_a.dot(vec_b);
+    }
+
+   // similarly, other methods
+};
+
+//template <typename T>
+class GPUBackend
+{
+public:
+ #ifdef HAVE_VIENNACL
```

Comment:
Remember to keep the header free of this dependency when you split your code into cpp and headers. Only the cpp file should have this if-def checks, if any.